src/cpp/utils/RefCountedPointer.hpp

```diff
@@ -0,0 +1,178 @@
+// Copyright 2024 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file RefCountedPointer.hpp
+ */
+
+#ifndef UTILS__REFCOUNTEDPOINTER_HPP
+#define UTILS__REFCOUNTEDPOINTER_HPP
+
+#include <atomic>
+#include <cassert>
+#include <condition_variable>
+#include <cstdint>
+#include <memory>
+#include <mutex>
+
+namespace eprosima {
+namespace fastdds {
+
+/**
+ * @brief Class to manage a local pointer with reference counting.
+ *
+ * It is similar to std::shared_ptr, but designed for cases where
+ * a shared pointer cannot be used due to API restrictions.
+ *
+ * USAGE:
+ * - On T class:
+ *   - Add a shared_ptr<RefCountedPointer<T>> local_ptr_ member.
+ *   - Call local_ptr_->deactivate() before destroying T.
+ *
+ * - On classes that need to use a pointer to T:
+ *   - Keep a copy of the shared_ptr<RefCountedPointer<T>>.
+ *   - Whenever you need to access T:
+ *     RefCountedPointer<T>::Instance instance(local_ptr_)
+ *     if (instance)
+ *     {
+ *         ptr->method();
+ *     }
+ */
+template<typename T>
+class RefCountedPointer
+{
+public:
+
+    class Instance;
+
+    explicit RefCountedPointer(
+            T* ptr)
+        : ptr_(ptr)
+        , is_active_(true)
+        , instances_(0)
+    {
+    }
```

Comment:
Add doxydoc here to explain that the pointer should not be `nullptr`, and add an `assert(nullptr != ptr);`

Reply:
Shouldn't we let `ptr` be `nullptr` for the cases in which `find_local_reader()` returns an `std::shared_ptr<LocalReaderPointer>(nullptr)` ?